### PR TITLE
run examples as part of tests

### LINKED ISSUE #624: Run examples as part of integration tests
this might help with #621

### PR DESCRIPTION
**Pull request by gr2m**
closes #624

### CODE REVIEW

**Comment by Ethan-Arrowood** [comment]
The pretest hook is failing due to `standard && standard-markdown` on a bunch of files within the examples directory. Any idea why it would be throwing errors now? Do I need to go through and make the changes manually?

**Comment by Ethan-Arrowood** [comment]
Had to add the `standard --fix` hook to pretest method. Also trying to work through a netrc error. Says "Error: ~/.netrc authentication type chosen but no credentials found for 'api.github.com'"
But we aren't even passing that host in the netrcAuth.js file.

**Comment by Ethan-Arrowood** [comment]
Okay so I got my local environment set up. Started working through what exists already. It seems as though a lot of the examples are broken throwing random errors all over the place. I'm not familiar with the entire API so I'm not sure the best way to settle each error. Furthermore, I tried to mimic the error handling inside `examples-test.js` but with no luck. Will continue to work on this and see where I can get it.

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
   },
   "devDependencies": {
     "@octokit/fixtures": "^3.0.0",
+    "bluebird": "^3.5.1",
```
Comment (gr2m) [inline package.json]: we will removed custom Promise support (see https://github.com/octokit/node-github/pull/623). You shouldn't need to add it back. If it's used in one of the examples, remove the `Promise` option instead

**Comment by gr2m** [comment]
make sure to rebase on the latest `master`, there have been a few changes, shouldn't affect examples though

File: package.json (hunk 1 above)
Comment (gr2m) [reply package.json]: I'd also remove the `testPromiseQ.js` file altogether

**Comment by gr2m** [comment]
pro tip :tm: while working on the examples, I'd run `node test/examples-test.js` instead of `npm test`, that will be faster and less noisy :) 

btw I like your `blacklist` workaround to get through all examples first, I can have a look in the one you have listed in there while you work on the rest. 

Thanks for helping Ethan, let me know if you have any other questions

File: examples/getRawBlob.js [hunk 2]
```diff
@@ -12,7 +10,7 @@ var github = new Client({
 
 github.authenticate({
   type: 'oauth',
-  token: testAuth['token']
+  token: 'add-your-real-token-here'
```
Comment (gr2m) [inline examples/getRawBlob.js]: for examples like this one where an auth token is not needed, I'd remove the `github.authenticate()` call altogether.

Also note that I changed `owner` and `repor` in all examples, so in this file you need to adapt the sha to point to one in the current repository.

**Comment by gr2m** [comment]
> "Error: ~/.netrc authentication type chosen but no credentials found for 'api.github.com'"

We would need to mock fs for return a mock `~/.netrc` file for that. My plan is to remove the built-in support for netrc anyway though, so let's just remove the example.

If tested it locally just to make sure it works as expected with the current code by created the file with this content

```
machine api.github.com
  login foo
  password bar
```

And I got a 401 as expected. So safe to remove it.

I've created an issue to remove netrc support: https://github.com/octokit/node-github/issues/633

**Comment by gr2m** [comment]
- [getPullRequestReviews.js](https://github.com/octokit/node-github/blob/72bff1e434ad3bbe0218558208f6ce402e109f2e/examples/getPullRequestReviews.js)  
why did you ignore this one? I would remove the empty line on top, change the owner and repo to `octokit` and `node-github`, remove the `github.authenticate()` call and the callback, otherwise I think this one is good to go :)
- [getPullRequestReview.js](https://github.com/octokit/node-github/blob/master/examples/getPullRequestReview.js)  
  We could just remove this one? If you want to keep it, I'd remove the `github.pullRequests.getReviews` call and keep only the `github.pullRequests.getReview` call, just hardcode a review `id` in the example. Also remove the authentication, callback and update the owner/repo like the other examples

**Comment by Ethan-Arrowood** [comment]
Okay so I rebased and started over because I realized a bunch of stuff.

First, a handful of these tests contain console.log(err, res[data]) statements that are making console outputs really gross when ran together. While most of the errors are caught by the examples-test.js file they are still being logged. As I work my through each example I shall replace it with console.log("Error thrown in . . .") so there is still an output just not something huge a unreadable.

Second I'm starting to delete things like the authenticate call as well as change the imports as you stated above.

Third following your instructions I began deleting tests such as the netrc one.

I have a busy weekend in front of me so I won't be able to commit much but expect more work next week 👍

File: examples/getContent.js [hunk 3]
```diff
@@ -11,5 +9,6 @@ github.repos.getContent({
   repo: 'node-github',
   path: ''
 }, function (err, res) {
-  console.log(err, res)
+  //console.log(err, res)
+  console.log("Error thrown in getContent.js");
```
Comment (gr2m) [inline examples/getContent.js]: you can remove the callback altogether :) See https://github.com/octokit/node-github/blob/624/run-examples-as-part-of-tests/examples/addCollaborator.js as an example

**Comment by gr2m** [comment]
cool great work @Ethan-Arrowood. No worries if you get to busy to continue to work on it, I can take it over any time and ind something else interesting for you to work on when you have more time

File: test/examples-test.js [hunk 4]
```diff
@@ -28,6 +38,12 @@ process.on('unhandledRejection', (error) => {
     return
   }
 
+  if (error.code === 403) {
+    // when API rate limit is reached 403 Forbidden is thrown
+    // Known files: addCollaborator.js, getContent.js
+    return
+  }
```
Comment (gr2m) [inline test/examples-test.js]: the rate limit error is actually a tricky one, I run into it myself today. I'll think about ways to work around it without adding more code to the examples

Comment (gr2m) [reply test/examples-test.js]: @Ethan-Arrowood I've pushed a change that should help with the rate limit errors, see 84dbe00. [Create a new token](https://github.com/settings/tokens/new), it does not need any access. Then create a `.env` file and put the token into it

```
GH_TOKEN=<your token here>
```

Then all requests from the examples are authenticated by default which bumps the rate limit for 60/minute to 6000/minute.

I'll need this finish up soon, if you can work on it today/tomorrow that's great, I can finish up the rest on Wednesday. Let me know if you have any questions :)

**Comment by Ethan-Arrowood** [comment]
@gr2m I hope what I'm pushing tonight knocks out a lot of the work for you. There are few more errors being thrown because some of the files have more complex error handlers and I wasn't sure what to do in those instances. Code coverage is up to 60% now (including all the example files). I should have some time tomorrow to review more and hopefully we can finish this up.

File: examples/getPullRequestReview.js [hunk 5]
```diff
@@ -1,32 +1,14 @@
...
-    repo: 'test',
-    number: 1,
-    id: reviewId
-  }, function (err, res) {
-    if (err) throw err
-    console.log(res)
-  })
+const reviewId = 'abcde'
+github.pullRequests.getReview({
+  owner: 'octoki',
```
Comment (gr2m) [inline examples/getPullRequestReview.js]: "t" is missing of octokit :)

File: examples/searchMostStarredRepos.js [hunk 6]
```diff
@@ -1,17 +1,9 @@
-
-
 const GitHubApi = require('github')
 
-
 const github = new GitHubApi({
     // debug: true
```
Comment (gr2m) [inline examples/searchMostStarredRepos.js]: I think we pass `debug: true` in all the other examples, maybe let's do it here, too?

File: examples/testPromiseGetNextPage.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 const GitHubApi = require('github')
 
-var gh = new GitHubApi();
+var gh = new GitHubApi()
```
Comment (gr2m) [inline examples/testPromiseGetNextPage.js]: can you also pass `{debug: true}` and rename `gh` to `github` to follow the other examples?

File: examples/getReleaseAsset.js [hunk 8]
```diff
@@ -51,9 +41,6 @@ github.repos.getReleases({
       // headers: {
       //     "Accept": "application/octet-stream"
       // }
```
Comment (gr2m) [inline examples/getReleaseAsset.js]: Let's remove the header entirely if it's not needed

File: examples/getStarred.js [hunk 9]
```diff
@@ -27,6 +19,7 @@ function getRepos (err, res) {
 }
 
 function outputStarredRepos () {
-  console.log(starredRepos.map(function (repo) { return repo['full_name'] }))
-  console.log('starred repos: ' + starredRepos.length)
+  // console.log(starredRepos.map(function (repo) { return repo['full_name'] }))
+  // console.log('starred repos: ' + starredRepos.length)
+
 }
```
Comment (gr2m) [inline examples/getStarred.js]: I would reduce the example to just `github.activity.getStarredRepos()` and leave pagination out it

File: examples/paginationCustomHeaders.js [hunk 10]
```diff
@@ -23,7 +15,6 @@ github.issues.getForRepo({
 }, function (err, res) {
```
Comment (gr2m) [inline examples/paginationCustomHeaders.js]: same as with `github.activity.getStarredRepos`, I would leave out pagination logic, let's just do `github.issues.getForRepo({...})`

File: examples/testPromiseGetNextPage.js [hunk 11]
```diff
@@ -1,8 +1,6 @@
-var Client = require('./../lib/index')
+const GitHubApi = require('github')
 
-var gh = new Client({
-  Promise: require('bluebird')
-})
+var gh = new GitHubApi()
```
Comment (gr2m) [inline examples/testPromiseGetNextPage.js]: rename `gh` to `github`, pass `{debug: true}` option like in the other examples

File: examples/updateReference.js [hunk 12]
```diff
@@ -1,23 +1,13 @@
...
-})
-
 github.gitdata.updateReference({
-  owner: 'kaizensoze',
-  repo: 'test2',
+  owner: 'octokit',
+  repo: 'node-github',
   ref: 'heads/master',
   sha: '81c559e2e8551982235bc86594cd86ffb135b053'
     // force: true
```
Comment (gr2m) [inline examples/updateReference.js]: fix indentation, remove 2 spaces beofer `//`

**Comment by Ethan-Arrowood** [comment]
@gr2m we are very close. Few more examples being buggy:
getNextPage
getReleaseAsset
searchMostStarredRepos
testPromiseGetNextPage

They all contain `if(err) throw err` lines, but I'm wary to just get rid of them. Let me know what you want me to do.

**Comment by gr2m** [comment]
@Ethan-Arrowood I think we are nearly there. There is one error about `required_pull_request_reviews`, I think this one is about branch protection, wanna look if you can figure it out?

**Comment by Ethan-Arrowood** [comment]
@gr2m yeah on it now :)

File: examples/testPromiseGetNextPage.js [hunk 13]
```diff
@@ -7,20 +7,21 @@ var github = new GitHubApi({
...
     .then(pager)
 }
 
-getAllOrgRepos('organization')
-  .then(function (orgRepos) {
-    console.log(orgRepos)
+getAllOrgRepos('github')
+  .then((orgRepos) => {
+    // orgRepos is array of all repositories meta data
   })
```
Comment (Ethan-Arrowood) [inline examples/testPromiseGetNextPage.js]: Based on the coverage tool, lines 11, 13, 14, 17 are all uncovered by the tests

File: examples/getReleaseAsset.js [hunk 14]
```diff
@@ -4,40 +4,41 @@ const github = new GitHubApi({
...
+  return github.repos.getAsset({
+    owner: 'octokit',
+    repo: 'node-github',
+    id: assetId
   })
 })
+
+.then(result => {
+  // result.data has asset properties
+})
```
Comment (Ethan-Arrowood) [inline examples/getReleaseAsset.js]: According to the code coverage tool we are missing a big chunk on this file: ... 29, 30, 34, 35. I believe the statement on line 12 is throwing an error which is skipping the remaining .then calls.

**Review by Ethan-Arrowood — COMMENTED** [review]
According to the code coverage tool we are missing a big chunk on this file: ... 29, 30, 34, 35. I believe the statement on line 12 is throwing an error which is skipping the remaining .then calls.

**Comment by gr2m** [comment]
don't worry about missing code coverage, we can set it to ignore files in the examples directory altogether, we are just interested in lines covered in the lib directory. I've made the setting via https://github.com/octokit/node-github/pull/627/commits/760a1145735cde961e8d18101f02b4ac8cece0dd

File: examples/getFollowing.js [hunk 15]
```diff
@@ -1,18 +1,7 @@
...
 
-
 const github = new GitHubApi({
   debug: true
 })
 
-github.authenticate({
-  type: 'oauth',
-  token: 'add-your-real-token-here'
-})
```
Comment (gr2m) [inline examples/getFollowing.js]: oh sorry I only now realised you removed these. We have to keep these in for all request that edit something, the invalid token here is important, the examples will fail and we expect them. But if we leave that out and I configure a token for my own account that has actual access to the `octokit/node-github` repository, I can end up changing stuff that I don't want to change :)

I'm sorry totally my fault I didn’t make this clear, I'll fix it

Comment (gr2m) [reply examples/getFollowing.js]: Lol I accidentally invited @defunkt to this repository because of that, sorry Chris :)

**Comment by gr2m** [comment]
@Ethan-Arrowood I think the pull request is all set, thanks a ton for your help! I've cleaned up the commits and made sure to set you as the author for the examples and the test files, is that okay with you? If you could give it a final review and approve the PR, that’d be great :)

Let me know if you are interested in contributing some more :) For example, we create fixtures for automated tests based on the example files and a few are still missing: https://github.com/octokit/fixtures/issues/14 It's quite interesting how the fixtures work :)

**Review by Ethan-Arrowood — APPROVED** [review]
LGTM 👍

**Comment by Ethan-Arrowood** [comment]
@gr2m this has been a lot of fun! Learned a lot and will definitely contribute more in the future. I'll have to take a few weeks off due to some upcoming interviews; however, I have a very big break from university coming up (2 full months) so I'll be back then to do more work 👍

**Comment by gr2m** [comment]
All the best for the upcoming interviews 🎉

**Pull request merged**